src/compiler/parser.ts

```diff
@@ -1537,6 +1537,159 @@ namespace ts {
             return false;
         }
 
+        const viableKeywordSuggestions = Object.keys(textToKeywordObj).filter(keyword => keyword.length > 2);
+
+        /**
+         * Attempts to consume the expected semicolon after an expression or property name.
+         *
+         * @param parentKind Kind of block the node and its expected semicolon are within.
+         * @param node Node preceding the expected semicolon location.
+         * @param initializer The preceding node's initializer, if it may exist and was able to parse.
+         * @remarks
+         * Use this to get a better error message than the generic "';' expected" if possible for
+         * known common variants of invalid syntax, such as mispelled names.
+         */
+        function parseSemicolonAfter(parentKind: SyntaxKind, node: Expression | PropertyName, initializer?: Node) {
+            // Consume the semicolon if it was explicitly provided.
+            if (canParseSemicolon()) {
+                if (token() === SyntaxKind.SemicolonToken) {
+                    nextToken();
+                }
+
+                return;
+            }
+
+            // If an initializer was parsed but there is still an error in finding the next semicolon,
+            // we generally know there was an error already reported in the initializer...
+            //   class Example { a = new Map([), ) }
+            //                                ~
+            if (initializer) {
+                // ...unless we've found the start of a block after a property declaration, in which
+                // case we can know that regardless of the initializer we should complain on the block.
+                //   class Example { a = 0 {} }
+                //                         ~
+                if (token() === SyntaxKind.OpenBraceToken && parentKind === SyntaxKind.PropertyDeclaration) {

```

Comment:
It's been bugging me that there's a whole block of hardcoded logic that only happens if `initializer` is provided _and_ the parent kind is a property declaration. I'll try extracting this out into its own function.

Reply:
Done in 7e49aa6.